File: python/robot-simulator/robot_simulator.py
# Globals for the directions
# Change the values as you see fit
EAST = {'offset': (1, 0)}
NORTH = {'offset': (0, 1)}
WEST = {'offset': (-1, 0)}
SOUTH = {'offset': (0, -1)}


class Robot(object):
    DIRECTIONS = [EAST, NORTH, WEST, SOUTH]

    def __init__(self, direction=NORTH, x=0, y=0):
        self.instruction_to_command = {
            'R': self.turn_right, 'L': self.turn_left, 'A': self.advance}

        self.direction = direction
        self.coordinates = (x, y)

    def advance(self):
        self.coordinates = tuple(
            map(sum, zip(self.coordinates, self.direction['offset'])))

    def turn_left(self):
        self.direction = self.DIRECTIONS[(self.DIRECTIONS.index(
            self.direction) + 1) % len(self.DIRECTIONS)]

    def turn_right(self):
        self.direction = self.DIRECTIONS[(self.DIRECTIONS.index(
            self.direction) - 1) % len(self.DIRECTIONS)]

    def move(self, instructions):
        for instruction in instructions:
            self.instruction_to_command[instruction]()

File: cpp/robot-simulator/robot_simulator.cpp
#include "robot_simulator.h"
#include <unordered_map>

namespace robot_simulator
{
const std::unordered_map<Bearing, std::pair<int, int>> BEARING_TO_OFFSET{
    {Bearing::NORTH, std::make_pair(0, 1)},
    {Bearing::EAST, std::make_pair(1, 0)},
    {Bearing::SOUTH, std::make_pair(0, -1)},
    {Bearing::WEST, std::make_pair(-1, 0)}};

Robot::Robot(const std::pair<int, int> &position, Bearing bearing) : position{position}, bearing{bearing} {}

const std::pair<int, int> &Robot::get_position() const { return position; }

Bearing Robot::get_bearing() const { return bearing; }

void Robot::advance()
{
    position.first += BEARING_TO_OFFSET.at(bearing).first;
    position.second += BEARING_TO_OFFSET.at(bearing).second;
}

void Robot::turn_left()
{
    bearing = static_cast<Bearing>((static_cast<int>(bearing) - 1 + static_cast<int>(Bearing::ITEM_COUNT)) % static_cast<int>(Bearing::ITEM_COUNT));
}

void Robot::turn_right()
{
    bearing = static_cast<Bearing>((static_cast<int>(bearing) + 1) % static_cast<int>(Bearing::ITEM_COUNT));
}

void Robot::execute_sequence(const std::string &instructions)
{
    for (char instruction : instructions)
    {
        if (instruction == 'A')
        {
            advance();
        }
        else if (instruction == 'L')
        {
            turn_left();
        }
        else if (instruction == 'R')
        {
            turn_right();
        }
    }
}
} // namespace robot_simulator
